libs/compute/include/hpx/compute/host/block_allocator.hpp

```diff
@@ -35,241 +36,242 @@
 #include <vector>
 
 namespace hpx { namespace compute { namespace host {
-    /// The block_allocator allocates blocks of memory evenly divided onto the
-    /// passed vector of targets. This is done by using first touch memory
-    /// placement. (maybe better methods will be used in the future...);
-    ///
-    /// This allocator can be used to write NUMA aware algorithms:
-    ///
-    /// typedef hpx::compute::host::block_allocator<int> allocator_type;
-    /// typedef hpx::compute::vector<int, allocator_type> vector_type;
-    ///
-    /// auto numa_nodes = hpx::compute::host::numa_domains();
-    /// std::size_t N = 2048;
-    /// vector_type v(N, allocator_type(numa_nodes));
-    ///
-    template <typename T,
-        typename Executor =
-            hpx::parallel::execution::restricted_thread_pool_executor>
-    struct block_allocator
-    {
-        typedef T value_type;
-        typedef T* pointer;
-        typedef const T* const_pointer;
-        typedef T& reference;
-        typedef T const& const_reference;
-        typedef std::size_t size_type;
-        typedef std::ptrdiff_t difference_type;
-
-        typedef Executor executor_type;
-
-        template <typename U>
-        struct rebind
+    namespace detail {
+        /// The policy_allocator allocates blocks of memory touched according to
+        /// the distribution policy of the given executor.
+        template <typename T, typename Policy,
+            typename Enable = typename std::enable_if<hpx::parallel::execution::
+                    is_execution_policy<Policy>::value>::type>
+        struct policy_allocator
         {
-            typedef block_allocator<U> other;
-        };
+            using policy_type = Policy;
 
-        typedef std::false_type is_always_equal;
-        typedef std::true_type propagate_on_container_move_assignment;
+            using value_type = T;
+            using pointer = T*;
+            using const_pointer = const T*;
+            using reference = T&;
+            using const_reference = T const&;
+            using size_type = std::size_t;
+            using difference_type = std::ptrdiff_t;
```

Comment:
Some of those are obsolete (as of C++20, see https://en.cppreference.com/w/cpp/memory/allocator). It's better to provide a specialization for `std::allocator_traits` instead (if needed at all - I don't think it's needed).

Reply:
Thanks, I learned something new! I'll remove the ones that aren't needed.

Reply:
I looked into this a bit more and I think we can keep (most of) these. What's deprecated are the typedefs for `std::allocator`, and they should presumably be accessed through `std::allocator_traits` instead. What doesn't seem to be deprecated is using typedefs for `value_type` etc. in a custom allocator. `std::allocator_traits` can be specialized, or it gets a particular typedef from the allocator itself (https://en.cppreference.com/w/cpp/memory/allocator_traits). Not having those typedefs in the allocator also means they're not available for use in the member function signatures. Using `std::allocator_traits` in the member function signatures requires the specialization again (otherwise it uses an incomplete type). On top of this the `hpx::compute::traits::allocator_traits` depends on the typedefs. That could be rewritten to use `std::allocator_traits` but I think that belongs in a separate PR.

All of this sounds complicated enough that I must be doing something wrong. Examples of specializing `std::allocator_traits` are scarce...

Reply:
Leaving the typedefs here is not a problem at all. The only reason I mentioned it was that we should not rely on those to be available in code referring to the allocator's properties. The best might be to make the deprecated typedefs `private`.

Reply:
Thanks, that makes sense. If it's ok with you I'd like to leave it like this here and open a separate PR to make `hpx::compute::traits::allocator_traits` not depend on the typedefs.